flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/sink/HBaseWriterAsyncHandler.java

```diff
@@ -0,0 +1,156 @@
+
+    private final Counter numRecordsOutErrorsCounter;
+
+    public HBaseWriterAsyncHandler(Counter numRecordsOutErrorsCounter) {
+        this.numRecordsOutErrorsCounter = numRecordsOutErrorsCounter;
+    }
+
+    /**
+     * For a given list of HBase write futures, this method will asynchronously analyze their
+     * result, and using the provided {@link ResultHandler}, it will instruct {@link
+     * org.apache.flink.connector.base.sink.writer.AsyncSinkWriter} to retry some mutations. In case
+     * of errors which should not be retried, the Flink job will stop with an error.
+     *
+     * @param futures list of HBase write futures
+     * @param processedMutationsInOrder list of mutations with their indices matching that of the
+     *     futures
+     * @param resultHandler result handler to manage retries and exceptions
+     */
+    public void handleWriteFutures(
+            List<CompletableFuture<Mutation>> futures,
+            List<SerializableMutation> processedMutationsInOrder,
+            ResultHandler<SerializableMutation> resultHandler) {
+        Preconditions.checkArgument(
+                futures.size() == processedMutationsInOrder.size(),
+                "Different number of HBase futures was supplied than mutations.");
+
+        ConcurrentLinkedQueue<FailedMutation> failedMutations = new ConcurrentLinkedQueue<>();
+
+        // Handle each future separately and store failures.
+        CompletableFuture<?>[] handledFutures = new CompletableFuture[futures.size()];
+        for (int i = 0; i < futures.size(); i++) {
+            final int index = i;
+            handledFutures[index] =
+                    futures.get(index)
+                            .exceptionally(
+                                    throwable -> {
+                                        failedMutations.add(
+                                                new FailedMutation(
+                                                        processedMutationsInOrder.get(index),
+                                                        throwable));
+                                        return null;
+                                    });
+        }
+
+        // Exceptions are already handled here, so it's safe to use `thenRun()`.
+        CompletableFuture.allOf(handledFutures)
+                .thenRun(
+                        () -> {
+                            handleFailedRequests(failedMutations, resultHandler);
+                        });
+    }
+
+    /**
+     * Handles mutations that failed to write to HBase.
+     *
+     * <p>This method increments the error counter and schedules the failed mutations for retry
+     * through the result handler. If the exception should not be retried, the job will fail instead
+     * with an exception.
+     *
+     * @param failedMutations the list of mutations that failed to write
+     * @param resultHandler the handler responsible for retry logic
+     */
+    private void handleFailedRequests(
```

Comment:
Can this get into an endless loop when some mutations always fail and we do not detect the given ex as fatal? :)